### PR TITLE
Adding support for JupyterLab

### PR DESCRIPTION
**Pull request by jacquekang**
Issue: https://github.com/lux-org/lux-widget/issues/16

To build the widget for jupyter-lab, run 
`pip install --upgrade jupyterlab`
`jupyter labextension update --all`
`jupyter labextension install .` in the lux-widget directory.

**Improvements:**
1) Double initialization bug doesn't happen in JupyterLab (not sure why but still a nice bonus) - toggling/rendering feels cleaner.
2) The frontend build is MUCH faster (8x speedup!) than for Jupyter Notebook - this should help with the frontend dev work
    i) Jupyter Notebook build time (via `sh install.sh`): ~4:03
    ii) JupyterLab build time (via `jupyter labextension install`): ~1:10
    iii) JupyterLab no minimize build time (via `jupyter lab build --minimize=False`): **~0:28**
For this issue https://github.com/lux-org/lux-widget/issues/34, perhaps we can note that we encourage anyone working on the frontend to use JupyterLab?

**Changes:**
1) Downgraded React dependency to 16.0.0 to be compatible with JupyterLab (and older versions).
2) Added "version" field in `package.json`, which is required to build the extension. 
3) Updated version to match the backend's version --> 0.2.0.
4) Fixed type issue in index.js that contributed to build failing.
5) Changed all module names to "luxwidget" - there was a conflict with "lux-widget" and "luxwidget" namings which caused the semver range bug - JupyterLab was inserting "lux-widget" as a key into some data structure, but using "luxwidget" as the query. Thus, no widget was registered.
6) I have Black automatically run on save so some files got reformatted + prettier extension for the typescript/js files.
7) Added documentation for both users and developers on how to install/build the JupyterLab extension.

### CODE REVIEW

File: src/plugin.ts [hunk 1]
```diff
@@ -12,43 +12,37 @@
...
+const luxPlugin: IPlugin<Application<Widget>, void> = ({
   id: EXTENSION_ID,
   requires: [IJupyterWidgetRegistry],
   activate: activateWidgetExtension,
-  autoStart: true
-};
-
-export default examplePlugin;
+  autoStart: true,
+});
```
Comment (cjachekang) [inline src/plugin.ts:35]:
```
as unknown) as IPlugin<Application<Widget>, void>;
// the "as unknown as ..." typecast above is solely to support JupyterLab 1
// and 2 in the same codebase and should be removed when we migrate to Lumino.
```
To fix the type issue on installation, you can add this - but ideally we would not have this in the code. It was there when we used phosphor but lumino should take care of this issue.

**Review by jrdzha — COMMENTED** [review]
Great job, Jake. Looking forward to the 28 sec compilte+install. Is there any chance you could add an install_lab.sh with those 3 install commands? Otherwise looks great!

**Comment by dorisjlee** [comment]
The changes in this PR looks great! The build time is much faster and the interactions are very smooth in JupyterLab. 
There are a couple of minor CSS issues:
1) The export alert box looks a bit weird probably because bootstrap CSS doesn't interact well with existing CSS in JupyterLab. The alert box also doesn't expire after X number of seconds.
image
2) The Toggle Button is slightly cut off at the very bottom. Minor adjustments to the Toggle button location should fix this.
</img></img>

Apart from the minor fixes, overall, this is really great progress and impressive work on this front!

**Review by dorisjlee — APPROVED** [review]

**Comment by cjachekang** [comment]
Notes:

1) One thing to note for the future is that JupyterLab doesn't currently support bootstrap, so a workaround is to copy-paste the CSS source code for whatever (in this case alerts) we need. 
2) Another is that if it's not much trouble, we should change the published name from "lux-widget" to "luxwidget" since this contributed to JupyterLab not working. It's changed in the codebase so if it gets changed automatically, then that's fine.

**Shipping Packages**

[Shipping package documentation](https://jupyterlab.readthedocs.io/en/stable/developer/extension_dev.html)
At the bottom under "Shipping Packages", looks like publishing to npm isn't necessary - should we do it anyway since Lux is on npm? Otherwise, we should just be able to build the widget after the `pip install` using `jupyter lab build`

**Comment by dorisjlee** [comment]
This looks good. I've made minor changes to the README.
1. Maybe in the future when we develop these in JupyterLab, we can find more native elements but the CSS fix looks good for now.
2.  I don't think we can change the lux-widget name for PyPI. What was the problem with JupyterLab with the name "luxwidget"?
3. Sure, let's change the version to 0.1.1 since that would be the next version for the release.

**Comment by cjachekang** [comment]
The problem wasn't the name necessarily but the fact that there existed both "lux-widget" and "luxwidget" in the frontend codebase so when JupyterLab inserted "luxwidget" into its data structure that held the widget and later queried that data structure using "lux-widget", it caused widget not found error. I think on install it should be fine though since in the codebase, I fixed this naming issue.

**Comment by dorisjlee** [comment]
Ok that sounds good. I'll merge this in then. Thanks!

**Comment by vvaezian** [comment]
> To build the widget for jupyter-lab, run 
> `pip install --upgrade jupyterlab` 
> `jupyter labextension update --all` 
> `jupyter labextension install .` in the lux-widget directory.

Regarding the last step, I didn't find any lux-widget, I found luxwidget though, so I ran the command there. But I get the error:
```
An error occurred.
ValueError: "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\luxwidget" is not a valid npm package
See the log file for details:  C:\Users\vahid\AppData\Local\Temp\jupyterlab-debug-7dr54gdg.log
```
Looking at the log file it says 
`npm ERR! Could not install from "..\..\..\Roaming\Python\Python310\site-packages\luxwidget" as it does not contain a package.json file.`

I looked into the subfolder `labextension`, it has a `package.json` file, so I ran the command in the subfolder, but I get
```
An error occurred.
ValueError: "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\luxwidget\labextension" is not a valid npm package
See the log file for details:  C:\Users\vahid\AppData\Local\Temp\jupyterlab-debug-ig2nezpy.log
```


<details><summary>logs</summary>

```
Node v12.16.1

Yarn configuration loaded.
> "C:\Program Files\nodejs\npm.CMD" pack C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\luxwidget\labextension

> luxwidget@0.1.11 prepare C:/Users/vahid/AppData/Roaming/Python/Python310/site-packages/luxwidget/labextension
> jlpm run clean && jlpm run build:prod

yarn run v1.21.1
$ jlpm run clean:lib
$ rimraf lib tsconfig.tsbuildinfo
Done in 1.44s.
yarn run v1.21.1
$ jlpm run build:lib && jlpm run build:labextension
$ tsc
Version 4.1.6
Syntax:   tsc [options] [file...]

Examples: tsc hello.ts
          tsc --outFile file.js file.ts
          tsc @args.txt
          tsc --build tsconfig.json

Options:
 -h, --help                                         Print this message.
 -w, --watch                                        Watch input files.
 --pretty                                           Stylize errors and messages using color and context (experimental).
 --all                                              Show all compiler options.
 -v, --version                                      Print the compiler's version.
 --init                                             Initializes a TypeScript project and creates a tsconfig.json file.
 -p FILE OR DIRECTORY, --project FILE OR DIRECTORY  Compile the project given the path to its configuration file, or to a folder with a 'tsconfig.json'.
 -b, --build                                        Build one or more projects and their dependencies, if out of date
 -t VERSION, --target VERSION                       Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'.
 -m KIND, --module KIND                             Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'.
 --lib                                              Specify library files to be included in the compilation.
                                                      'es5' 'es6' 'es2015' 'es7' 'es2016' 'es2017' 'es2018' 'es2019' 'es2020' 'esnext' 'dom' 'dom.iterable' 'webworker' 'webworker.importscripts' 'webworker.iterable' 'scripthost' 'es2015.core' 'es2015.collection' 'es2015.generator' 'es2015.iterable' 'es2015.promise' 'es2015.proxy' 'es2015.reflect' 'es2015.symbol' 'es2015.symbol.wellknown' 'es2016.array.include' 'es2017.object' 'es2017.sharedmemory' 'es2017.string' 'es2017.intl' 'es2017.typedarrays' 'es2018.asyncgenerator' 'es2018.asynciterable' 'es2018.intl' 'es2018.promise' 'es2018.regexp' 'es2019.array' 'es2019.object' 'es2019.string' 'es2019.symbol' 'es2020.bigint' 'es2020.promise' 'es2020.sharedmemory' 'es2020.string' 'es2020.symbol.wellknown' 'es2020.intl' 'esnext.array' 'esnext.symbol' 'esnext.asynciterable' 'esnext.intl' 'esnext.bigint' 'esnext.string' 'esnext.promise' 'esnext.weakref' 
 --allowJs                                          Allow javascript files to be compiled.
 --jsx KIND                                         Specify JSX code generation: 'preserve', 'react-native', or 'react'.
 -d, --declaration                                  Generates corresponding '.d.ts' file.
 --declarationMap                                   Generates a sourcemap for each corresponding '.d.ts' file.
 --sourceMap                                        Generates corresponding '.map' file.
 --outFile FILE                                     Concatenate and emit output to single file.
 --outDir DIRECTORY                                 Redirect output structure to the directory.
 --removeComments                                   Do not emit comments to output.
 --noEmit                                           Do not emit outputs.
 --strict                                           Enable all strict type-checking options.
 --noImplicitAny                                    Raise error on expressions and declarations with an implied 'any' type.
 --strictNullChecks                                 Enable strict null checks.
 --strictFunctionTypes                              Enable strict checking of function types.
 --strictBindCallApply                              Enable strict 'bind', 'call', and 'apply' methods on functions.
 --strictPropertyInitialization                     Enable strict checking of property initialization in classes.
 --noImplicitThis                                   Raise error on 'this' expressions with an implied 'any' type.
 --alwaysStrict                                     Parse in strict mode and emit "use strict" for each source file.
 --noUnusedLocals                                   Report errors on unused locals.
 --noUnusedParameters                               Report errors on unused parameters.
 --noImplicitReturns                                Report error when not all code paths in function return a value.
 --noFallthroughCasesInSwitch                       Report errors for fallthrough cases in switch statement.
 --types                                            Type declaration files to be included in compilation.
 --esModuleInterop                                  Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'.
 @<file>                                            Insert command line options and files from a file.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! luxwidget@0.1.11 prepare: `jlpm run clean && jlpm run build:prod`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the luxwidget@0.1.11 prepare script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\vahid\AppData\Roaming\npm-cache\_logs\2022-05-14T18_05_55_020Z-debug.log

Traceback (most recent call last):

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\debuglog.py", line 47, in debug_logging
    yield

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\labextensions.py", line 144, in start
    ans = self.run_task()

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\labextensions.py", line 191, in run_task
    [

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\labextensions.py", line 192, in <listcomp>
    install_extension(

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\commands.py", line 421, in install_extension
    return handler.install_extension(extension, pin=pin)

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\commands.py", line 645, in install_extension
    info = self._install_extension(extension, tempdir, pin=pin)

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\commands.py", line 1710, in _install_extension
    info = self._extract_package(extension, tempdir, pin=pin)

  File "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\jupyterlab\commands.py", line 1783, in _extract_package
    raise ValueError(msg % source)

ValueError: "C:\Users\vahid\AppData\Roaming\Python\Python310\site-packages\luxwidget\labextension" is not a valid npm package

Exiting application: lab
```

</details>


@cjachekang Where exactly should the command in the third step be run?

**Pull request merged**